README.md
mrCaNaVaR analyzes the mapping read depth to discover large segmental duplications and deletions. It also has the capability of predicting absolute copy numbers of genomic intervals.

# Citation

* Personalized copy number and segmental duplication maps using next-generation sequencing. Can Alkan, Jeffrey M. Kidd, Tomas Marques-Bonet, Gozde Aksay, Francesca Antonacci, Fereydoun Hormozdiari, Jacob O. Kitzman, Carl Baker, Maika Malig, Onur Mutlu, S. Cenk Sahinalp, Richard A. Gibbs, Evan E. Eichler. Nature Genetics, Oct, 41(10):1061-1067, 2009.
* Whole-genome shotgun sequence CNV detection using read depth. Fatma Kahveci and Can Alkan. Methods Mol Biol., 1833:61-72, 2018.

# Basics & Definitions

## Pronunciation: 

mrCaNaVaR is pronounced as /mɪstər ʤanavar/, which sounds like "mister juh-nuh-vuhR".

## Requirements:

* mrFAST or mrsFAST for read mapping stage. No other mapper is recommended to use with mrCaNaVaR for characterization of duplications and absolute copy number prediction. Deletion discovery *might* work, but other mappers are untested.
* zlib for the ability to read compressed SAM files.
* C compiler (mrCaNaVaR is developed with gcc version 4.1.2)

Please download the latest version from our download page and then unzip the downloaded file. Run 'make' to build mrFAST.

## Coordinate space:

mrCaNaVaR uses BED-style coordinates for genomic locations (chrom/chromStart/chromEnd) to facilitate easier integration with genome browsers and tools like BEDtools.  The first base in a chromosome is numbered 0. The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart=0, chromEnd=100, and span the bases numbered 0-99.

## Sliding windows: 

mrCaNaVaR operates on three classses of sliding windows over the reference genome to estimate absolute copy numbers and discover duplicated and deleted regions. These three windows are:

* *Long Windows (LW)* are used to find locations of deletions and duplications. The LWs are overlapping. By default the window size size is 5000 bp of non-masked characters (i.e. A, C, G, T only), and the slide size is 1000 bp of any characters. For example, in repeat masked NCBI Build 35, the first two LWs are:                   

          chr1    0       6792
          chr1    1000    6792
          chr1    2000    7203

          Here, the first window has 6792 any characters, but 5000 non-N characters (thus 1792 N characters). 
          We then slide the window by 1000 "any" base pairs, starting the second window at position 1000.  
          The end of the second window is again 6792 since all of the first 1000 characters of chr1 is masked. 
          The third window starts at position 2000 and contains 5203 any characters, where 5000 are non-N.

* *Short Windows (SW)* are used to refine breakpoints of detected duplication and deletion intervals. Default window size is 1000 bp of non-masked characters, and the slide size is 1000 bp of any characters, similar to LW.

* *Copy Windows (CW)* are used to predict absolute copy numbers. CWs are non-overlapping windows. Default window size is 1000 bp of non-masked characters.

# Before using mrCaNaVaR

* Mask the repeats in your reference genome assembly. For the human genome assemblies (NCBI Build 35, 36, GRCh37) we recommend both:
* RepeatMasker with sensitive "-s" parameter enabled. The repeat masked versions of many genome assemblies are available at the UCSC Genome Browser. 
* Tandem Repeats Finder  (parameters: trf input_file  2 7 7 80 10 50 500 -m)
* The repeats should be masked with N characters in your genome. Lowercase letters will be treated as non-repeats.

# Read Mapping:

* Please use either mrFAST or mrsFAST in single-read mapping mode (not paired-end). Recommended edit/hamming distance threshold is 5% of the read length (i.e. 2 for 36 bp reads, 4 for 76 bp reads, etc.). Using other aligners might be possible if all map locations (including suboptimal alignments) are reported, however, we strongly recommend mrFAST or mrsFAST that were specifically developed for SV/CNV analysis.

* Use the same repeat masked version of the reference genome described above for building the mrFAST/mrsFAST index and read mapping.

# Using mrCaNaVaR:

* Mask the common and tandem repeats in your reference genome assembly of choice. Replace repeat characters with N.
* Build mrFAST or mrsFAST search index using the same genome.
* Map your reads with mrFAST or mrsFAST.
* Prepare genome configuration file with mrCaNaVaR (PREP mode; see below).
* Calculate read depth values from the mapping information generated by mrFAST or mrsFAST using mrCaNaVaR (READ mode; see below).
* Call segmental duplication and deletion intervals, and predict absolute copy numbers using mrCaNaVaR (CALL mode; see below).

# Run Modes

**mrCaNaVaR has three run modes**

* **PREP mode (--prep)** is used to prepare a reference genome for analysis. Users need to run mrCaNaVaR in this mode only once for a new reference genome assembly (or versions of it with different repeat masking or different window sizes). In this mode mrCaNaVaR will:

    * Load reference genome file (FASTA format), calculate the coordinates and GC content of three window classes (LW, SW, and CW).
    * Save this information in a "configuration" file.

    *Required files and parameters:*

    * *fasta: Repeat masked reference genome in FASTA format (e.g. hg17_masked.fa)*
    * *gaps: Coordinates of assembly gaps in the reference genome in BED format (e.g. hg19.gaps.bed)*

    *Output:*

    * *conf: Reference configuration file. This file is in a native binary format. (e.g. hg17.cnvr)*

    *Sample command line:*

		mrcanavar --prep -fasta hg17_masked.fa -gaps hg17.gaps.bed -conf hg17.cnvr

    *Optional parameters:*

		-lw_size   <lw_size>  : Long window span size. Default is 5000.
		-lw_slide  <lw_slide> : Long window slide size. Default is 1000.
		-sw_size   <sw_size>  : Short window span size. Default is 1000.
		-sw_slide  <sw_slide> : Short window slide size. Default is 1000.
		-cw_size   <cw_size>  : Copy number window size. Default is 1000.
		-pseudoa   <bed_file> : Coordinates for pseudoautosomal regions in the 
                		        reference genome in BED format.


* **READ mode (--read)** is used after read mapping stage to load mapping information and calculate read depths over the LW, SW, and CW intervals. The mapping information should be in SAM format; either uncompressed (*.sam), or  compressed with gzip (*.sam.gz). Both mrFAST and mrsFAST are capable of generating map files in plain SAM format and gzip-compressed SAM. In this mode mrCaNaVaR will:   

    * Load reference configuration file.
    * Scan a given directory for *.sam or *.sam.gz files, and calculate the read depth in three window classes.
    * Save the read depth information in a depth output file in a native binary format. In addition it will generate three human-readable text filesfor read depths and GC contents of LW, SW, and CW  intervals.
    * Perform GC correction of the read depth values over the LW, SW, and CW intervals. This is new due to the CONC mode and multiple library support.
    
    *Required files:*
    * *conf: Reference configuration file created using the PREP mode.*
    * *samdir: Plain SAM files (*.sam) or gzip-compressed SAM files (*.sam.gz) in a directory (e.g. sam_files_are_here/).*
    
    *Output:*
    * *depth: Read depth file in a native binary format.*
    * *Three human readable text files that contain the same information as the file above. These three files are for manual inspection purposes only, if needed, and will not be loaded and needed by mrCaNaVaR.*
    
    *Sample command line:*
    
		mrcanavar --read -conf hg17.cnvr -samdir sam_files_are_here/ -depth mysample.depth

    *Optional parameters:*

		--gz                    : Indicates the SAM files are compressed in gzip format.
		-nsam <first> <second>  : Choose a sequence of SAM files between two indexes.


* **CONC mode (--conc)** is used to merge >=2 *.depth files created using the READ mode. The CONC mode is optional, and needed only if the SAM files are read into a number of *.depth files; or, if there are more than one library preparation (which should be read into different *.depth files). In this mode, mrCaNaVaR will:

    * Load reference configuration file.
    * Load read depth files through -concdepth parameter.
    * Merge depth files and output the merged LW, SW, and CW intervals.

    *Required files:*
    * *conf: Reference configuration file created using the PREP mode.*
    * *concdepth: List of read depth files to merge.*
    
    *Output:*
    * *depth: Read depth file in a native binary format.*
    * *Three human readable text files that contain the same information as the file above. These three files are for manual inspection purposesonly, if needed, and will not be loaded and needed by mrCaNaVaR.*
    
    *Sample command line:*

		mrcanavar --conc -conf hg17.cnvr -concdepth lib1.depth lib2.depth lib3.depth -depth merged.depth

* **CALL mode (--call)** is the final step in CNV analysis using mrCaNaVaR. In this mode, mrCaNaVaR will:

    * Load reference configuration file.
    * Load read depth file.
    * Perform GC correction of the read depth values over the LW, SW, and CW intervals if the *.depth file is created with mrCaNaVaR version earlier than 0.5.
    * Predict absolute copy number over the CW intervals.
    * Predict duplicated and deleted regions in the analyzed sample using the LW and SW intervals.

    *Required files:*
    * *conf: Reference configuration file created using the PREP mode.*
    * *depth: Read depth file created using the READ mode.*

    *Output:*
    * *o: Output file prefix (out_prefix).*
    
    *Output files:*
    * *out_prefix.log:  Log file that describes what mrCaNaVaR did. Includes read depth statistics.*
    * *out_prefix.cw_norm.bed: GC-corrected read depth values for the inferred control regions for CW intervals.*
    * *out_prefix.copynumber.bed: Estimated copy numbers over CW intervals.*

    *Sample command line:*

		mrcanavar --call -conf hg17.cnvr -depth mysample.depth -o mysample

    *Optional parameters:*

		--xx : Set gender of the sequenced sample as female. Mammalian genomes only. 
					Default is autodetect.
		--xy : Set gender of the sequenced sample as male. Mammalian genomes only. 
					Default is autodetect.
		--multgc : Perform multiplicative GC correction. Default is additive.
		-mindup <min_dup_len>   : Minimum duplication length. Default is 10000.
		-gene <genelist>        : Coordinates for genes for calculating gene-based copy numbers.
		-cont_win <cwin_number> : Contiguous window number to look for high/low depth windows. 
					  Default is 7.
		-cut_win  <cwin_cutoff> : Window number cutoff. Default is 6.
		--verbose : Verbose output.

# Automated run.

To compile:
	
	make && make auto

mrcanavar-auto is a wrapper for mrsFAST and mrCaNaVaR. It assumes that both mrsFAST and mrCaNaVaR are available in your PATH. It also assumes that the mrsFAST index file is available in the same directory as the reference genome. Simply run:

	mrcanavar-auto --ref repeat_masked_ref.fa --input file1.fastq.gz,file2.fastq.gz,file3.fastq.gz --conf repeat_masked_ref.cnvr --gene repeat_masked_ref.genes.bed

Full parameter list:
	
	--input [list]: comma-separated FASTQ(.gz) files.
	--ref [ref.fa]: repeat masked reference genome. mrsFAST index (ref.fa.index) should be in the same directory.
	--conf [ref.cnvr]: mrCaNaVaR config file.
	--gene [genes.bed]: List of genes.
	--threads [int]: number of threads for mrsFAST.
	--skip-mapping: Skip mrsFAST mapping. Use this only if you have the mapping output and rerunning mrCaNaVaR for some reason.
	--cloud: Cloud mode, the directory info from the input file names will be stripped for output file generation.
	--dry-run: Do not execute commands, just print them.



Changelog
December 19, 2018:
	version 1.1.3:
		* A corner case bug is fixed.

November 30, 2018:
	version 1.1.2:
		* Memory leak in gene copy number calculation is fixed.

November 28, 2018:
	version 1.1.1:
		* Dry run and "cloud mode" added to mrcanavar-auto.

November 27, 2018:
	version 1.1:
		* mrcanavar-auto added for automated mapping & calling. Assumes that some files are available in the working directory. Use it with caution.

November 20, 2018:
	version 1.0:
		* SAM(.gz) files in READ mode can now be passed as comma-separated list.

April 18, 2014:
	version 1.0 beta:
		* GC normalization is now less strict
		* Duplication windows are calculated through LW and SW. CW info for further filtering to be added.

September 4, 2013:
	version 0.51 released:
		* No new updates, only the bugfix below.

July 15, 2013:
		* Minor bug in CW counting is fixed.

July 2, 2013:
	version 0.5:
		* SAM files can now separately be read into *.depth files, and then merged. 
		* -nsam parameter to read specific SAM files in the READ mode.
		* New CONC mode (--conc) to concatenate/merge different *.depth files using the new -concdepth parameter. These different files may be from different library preparations (see below).
		* Multiple library support added.
		* Several bug fixes.

November 26, 2012:
	version 0.41:
		 * Bugs in non-autosomal chromosome processing for male samples are fixed.

September 23, 2012:
	version 0.4:
		 * Pseudoautosomal support added through -pseudoa paramater. You can enter the pseudoautosomal regions in BED format in the PREP mode. This is optional; but the configuration file generated and used by this version is not compatible with the previous versions. You need to recreate the configuration file using the PREP mode.
		 * Code cleanup to facilitate easier addition of new features.

June 1, 2012
	version 0.35:
		 * SAM headers are now ignored.

December 7, 2011
	version 0.34:
		 * Segmentation fault with some kernels & gcc due to an unallocated string is fixed.

October 13, 2011
	version 0.33:
		 * Compilation errors in MacOSX are fixed.
    		 * Compilation error in x86_64 systems with gcc version > 4.33 is fixed.

July 22, 2011
	version 0.32:
		* Bug in parsing chromosome names that end with whitespace is fixed.

June 22, 2011
	version 0.31:
		* Bug causing infinite loop in control region selection is fixed.

April 16, 2011
	version 0.3:
		* Bug in GC normalization fixed
		* Bug in CN estimation in chrX and chrY for male mammals is fixed.

March 3, 2011
	GC counting bug in LW when an assembly gap is 1 character is fixed. (version 0.22)

March 2, 2011
	Bug causing segmentation fault when CW_SIZE is set to a small value is fixed.  (version 0.21)

February 25, 2011
	A bug in GC normalization is fixed. (version 0.2)

January 23, 2011
	SW normalization implemented  (copy paste from LW ;-) )

December 27, 2010
	LW normalization implemented

December 21, 2010
	a bug in handling file names is fixed

December 14, 2010
	fixes to control selection: normalization is done AFTER controls are selected
	Multiplicative GC scaling option (--multgc)

November 18, 2010
	added removing low coverage windows from controls

November 14, 2010
	version 0.1-rc
	added iteration until min >= mean - 3std ; and 2std for chrX

November 11, 2010
	version 0.1-alpha


mrcanavar.c
#include "mrcanavar.h"



int main(int argc, char **argv){

  int i;
  int ctr;
  int fileCount;
  int fileEnd;
  int fileStart;

  char gzSAM;

  char *indirSAM;
  char *inSAMlist;
  char *depthFile;
  char *out_prefix;
  char *gene;
  char **c_depths;

  init_globals();

  gzSAM      = 0;
  indirSAM   = NULL;
  inSAMlist  = NULL;
  depthFile  = NULL;
  out_prefix = NULL;
  gene=NULL;
  n_sam_e = 0;
  n_sam_b =0;
  fileCount = 0;
  fileEnd = 0; 
  fileStart = 1;
  
  c_depths = NULL;

  for (i=1; i<argc; i++){


    /* modes */
    if (!strcmp(argv[i], "--prep"))
      set_runmode(PREP);
    else if (!strcmp(argv[i], "--read"))
      set_runmode(READSAM);
    else if (!strcmp(argv[i], "--call"))
      set_runmode(CALL);
    else if (!strcmp(argv[i], "--conc"))
      set_runmode(CONC);

    /* compulsory parameters for all modes */
    else if (!strcmp(argv[i], "-conf"))
      set_str(&GENOME_CONF, argv[i+1]);

    /* compulsory parameters for the PREP mode */
    else if (!strcmp(argv[i], "-fasta"))
      set_str(&GENOME_FASTA, argv[i+1]);
    else if (!strcmp(argv[i], "-gaps"))
      set_str(&GENOME_GAPS, argv[i+1]);

    /* optional parameters for the PREP mode */
    else if (!strcmp(argv[i], "-pseudoa"))
      set_str(&GENOME_PSEUDO , argv[i+1]);
    else if (!strcmp(argv[i], "-lw_size"))
      LW_SIZE = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-sw_size"))
      SW_SIZE = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-cw_size"))
      CW_SIZE = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-lw_slide"))
      LW_SLIDE = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-sw_slide"))
      SW_SLIDE = atoi(argv[i+1]);


    /* compulsory parameters for  READ, CALL and CONC modes */
    else if (!strcmp(argv[i], "-depth"))
      set_str(&depthFile, argv[i+1]);
    
    /* compulsory paramaters for CONC mode*/
    else if (!strcmp(argv[i], "-concdepth")){
      
      fileStart=i+1; fileEnd=i+1; i++;

      while (i<argc && argv[i][0]!='-')
	fileEnd=i++;
      i--;

      fileCount = fileEnd-fileStart+1;

      if (fileEnd == fileStart ){
	fprintf(stderr, "No files are given.\n");
	return 0;
      }

      c_depths = getMem( fileCount * sizeof * c_depths);

      for ( ctr=0; ctr<fileCount; ctr++){
	set_str(&c_depths[ctr], argv[ctr+fileStart]);
      }

    }	 
      
 

    /* compulsory parameters for the READ mode */

    else if (!strcmp(argv[i], "-samdir"))
      set_str(&indirSAM, argv[i+1]);
  
    else if (!strcmp(argv[i], "-samlist"))
      set_str(&inSAMlist, argv[i+1]);
  
  
    /* optional parameters for the SAM mode */
    else if (!strcmp(argv[i], "--gz"))
      gzSAM = 1;
    /* optional parameters for the SAM mode */
    else if (!strcmp(argv[i], "-nsam")){
      n_sam_b = atoi(argv[i+1]);
      n_sam_e = atoi(argv[i+2]);
    }
    /* compulsory parameters for the CALL mode */
    else if (!strcmp(argv[i], "-o"))
      set_str(&out_prefix, argv[i+1]);

    /* optional parameters for the CALL mode */
    else if (!strcmp(argv[i], "-cont_win"))
      CONT_WINDOW = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-cut_win"))
      CUT_WINDOW = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "--xx"))
      set_gender(FEMALE);
    else if (!strcmp(argv[i], "--xy"))
      set_gender(MALE);
    else if (!strcmp(argv[i], "--multgc"))
      MULTGC = 1;
    else if (!strcmp(argv[i], "-mindup"))
      MIN_DUP = atoi(argv[i+1]);
    else if (!strcmp(argv[i], "-gene"))
      set_str(&gene, argv[i+1]);


    /* generic stuff */
    else if (!strcmp(argv[i], "-v")){
      fprintf (stdout, "\nmrCaNaVaR version %s\nLast update: %s\n", VERSION, LAST_UPDATE);
      return 0;
    }
    else if (!strcmp(argv[i], "-h")){
      printHelp(argv[0]);
      return 0;
    }
    else if (!strcmp(argv[i], "--verbose"))
      VERBOSE = 1;
    else if (!strcmp(argv[i], "--disable-sam-check"))
      CHECKSAM = 0;
  }


  switch(RUNMODE){
  case NONE:
    print_error("No run mode selected.\nSelect mode: --prep, ---read,--call or --conc\n");
    break;
  case PREP:
    fprintf(stdout, "Mode: Prepare genome...\n");
    prep_genome();
    break;
  case READSAM:
    fprintf(stdout, "Mode: Read SAM ...\n");

    if (indirSAM == NULL && inSAMlist == NULL){
      print_error("Use either -samdir or -samlist parameter to pass the input SAM files.\n");
    }
    
    else if (indirSAM != NULL && inSAMlist != NULL){
      print_error("Use either -samdir or -samlist parameter to pass the input SAM files. Not both!\n");
    }
    
  
    if (indirSAM != NULL)
      read_mapfiles(indirSAM, depthFile, gzSAM, SAMDIR);
    else
      read_mapfiles(inSAMlist, depthFile, gzSAM, SAMLIST);
    break;
  case CALL:
    fprintf(stdout, "Mode: Call copy numbers ...\n");
    call_cnv(depthFile, out_prefix,gene);
    break;
  case CONC:
    fprintf(stdout, "Mode: Merge depth files...\n");
    conc_depth(c_depths, fileCount, depthFile);
    break;

  default:
    break;
  }

  return 0;

}


void printHelp(char *binfile){

  fprintf (stdout, "\nmrCaNaVaR version %s\nLast update: %s\n", VERSION, LAST_UPDATE);
  fprintf (stdout, "%s --<mode>  [options] \n\n", binfile);

  fprintf (stdout, "========            RUN MODES            ========\n\n");

  fprintf (stdout, "\t--prep                  : Prepare reference genome configuration file.\n");
  fprintf (stdout, "\t--read                  : Read mapping information from SAM files.\n");
  fprintf (stdout, "\t--conc                  : Merge .depth files that are given as parameters (optional mode).\n");
  fprintf (stdout, "\t--call                  : Call CNVs and predict copy numbers.\n\n");

  fprintf (stdout, "======== PREP MODE COMPULSORY PARAMETERS ========\n\n");
  fprintf (stdout, "\t-fasta <fasta_file>     : FASTA file for the reference genome.\n");
  fprintf (stdout, "\t-gaps  <gaps_file>      : Gap coordinates of the reference genome in BED format.\n");
  fprintf (stdout, "\t-conf  <config_file>    : Reference configuration file (output).\n\n");

  fprintf (stdout, "========  PREP MODE OPTIONAL PARAMETERS  ========\n\n");
  fprintf (stdout, "\t-lw_size   <lw_size>    : Long window span size. Default is 5000.\n");
  fprintf (stdout, "\t-lw_slide  <lw_slide>   : Long window slide size. Default is 1000.\n");
  fprintf (stdout, "\t-sw_size   <sw_size>    : Short window span size. Default is 1000.\n");
  fprintf (stdout, "\t-sw_slide  <sw_slide>   : Short window slide size. Default is 1000.\n");
  fprintf (stdout, "\t-cw_size   <cw_size>    : Copy number window size. Default is 1000.\n");
  fprintf (stdout, "\t-pseudoa   <bed_file>   : Coordinates for pseudoautosomal regions in the reference genome in BED format.\n\n");

  fprintf (stdout, "======== READ MODE COMPULSORY PARAMETERS ========\n\n");
  fprintf (stdout, "\t-conf   <config_file>   : Reference configuration file (input).\n");
  fprintf (stdout, "\t-samdir <sam_dir>       : Directory that contains SAM files for mapping information.\n");
  fprintf (stdout, "\t-samlist <sam_list>     : Alternative to -samdir. Comma separated list of SAM files for mapping information.\n");
  fprintf (stdout, "\t-depth  <depth_file>    : Read depth file (output).\n\n");
  fprintf (stdout, "========  READ MODE OPTIONAL PARAMETERS  ========\n\n");
  fprintf (stdout, "\t--gz                    : Indicates that the SAM files are compressed in gzip format.\n");
  fprintf (stdout, "\t-nsam <first> <second>  : Choose a sequence of SAM files between two indexes.\n\n");

  fprintf (stdout, "======== CONC MODE COMPULSORY PARAMETERS ========\n\n");
  fprintf (stdout, "\t-conf     <config_file> : Reference configuration file (input).\n");
  fprintf (stdout, "\t-concdepth <depth_file> : Read depth file(s).\n");
  fprintf (stdout, "\t-depth     <depth_file> : Read depth file (output).\n\n");

  fprintf (stdout, "======== CALL MODE COMPULSORY PARAMETERS ========\n\n");
  fprintf (stdout, "\t-conf   <config_file>   : Reference configuration file (input).\n");
  fprintf (stdout, "\t-depth  <depth_file>    : Read depth file (input).\n");
  fprintf (stdout, "\t-o      <out_prefix>    : Prefix for the output file names.\n\n");

  fprintf (stdout, "========  CALL MODE OPTIONAL PARAMETERS  ========\n\n");
  fprintf (stdout, "\t--xx                    : Set gender of the sequenced sample as female. Mammalian genomes only. Default is autodetect.\n");
  fprintf (stdout, "\t--xy                    : Set gender of the sequenced sample as male. Mammalian genomes only. Default is autodetect.\n");
  fprintf (stdout, "\t--multgc                : Perform multiplicative GC correction. Default is additive.\n");
  fprintf (stdout, "\t-mindup <min_dup_len>   : Minimum duplication length. Default is 10000.\n");
  fprintf (stdout, "\t-gene <genelist>        : Coordinates for genes for calculating gene-based copy numbers.\n");
  fprintf (stdout, "\t-cont_win <cwin_number> : Contiguous window number to look for high/low depth windows. Default is 7.\n");
  fprintf (stdout, "\t-cut_win  <cwin_cutoff> : Window number cutoff. Default is 6.\n");
  fprintf (stdout, "\t--verbose               : Verbose output.\n\n");


}
